### ISSUE TITLE
Invalid SQL Error when using Class.with_role with unsaved record

### ISSUE THREAD

**Issue by chriskaukis** [RolifyCommunity/rolify#382]
I feel like I might be doing something wrong, but here is the behavior I am experiencing and how to reproduce. In summary, when using Class.with_role(:role_name, user) (for example) on a new user record (user.persisted? is false, user.new_record? is true) it throws a SQL error. In my case it's a PG::SyntaxError.  Here is the output of the ActiveRecord Relation / SQL:

```
[24] pry(main)> user = User.new
[25] pry(main)> Thing.with_role(:admin, user)
  Thing Load (1.7ms)  SELECT "things".* FROM "things" INNER JOIN "roles" ON "roles".resource_type IN ('Thing') AND
                                    ("roles".resource_id IS NULL OR "roles".resource_id = "things".id) WHERE ("roles".name IN ('admin') AND "roles".resource_type IN ('Thing')) AND ("roles".id IN () AND ((resource_id = "things".id) OR (resource_id IS NULL)))
=> #<Thing::ActiveRecord_Relation:0x3fc1d097eac4>
```

Note: `"roles".id IN ()` portion.

We are using CanCanCan, thus if we do something like this in our ability class:

```
  def initialize(user)
    user ||= User.new

    if user.has_role? :admin
      can :manage, :all
    end

    can :manage, Thing, id: Thing.with_role(:admin, user).pluck(:id)
  end
```

It throws a PG::SyntaxError:

```
[26] pry(main)> Thing.with_role(:admin, user).pluck(:id)
   (1.1ms)  SELECT "things"."id" FROM "things" INNER JOIN "roles" ON "roles".resource_type IN ('Thing') AND
                                    ("roles".resource_id IS NULL OR "roles".resource_id = "things".id) WHERE ("roles".name IN ('admin') AND "roles".resource_type IN ('Thing')) AND ("roles".id IN () AND ((resource_id = "things".id) OR (resource_id IS NULL)))
ActiveRecord::StatementInvalid: PG::SyntaxError: ERROR:  syntax error at or near ")"
LINE 2: ...esource_type IN ('Thing')) AND ("roles".id IN () AND ((re...
                                                             ^
: SELECT "things"."id" FROM "things" INNER JOIN "roles" ON "roles".resource_type IN ('Thing') AND
                                    ("roles".resource_id IS NULL OR "roles".resource_id = "things".id) WHERE ("roles".name IN ('admin') AND "roles".resource_type IN ('Thing')) AND ("roles".id IN () AND ((resource_id = "things".id) OR (resource_id IS NULL)))
from /Users/chris/.rvm/gems/ruby-2.2.2@myproject/gems/rack-mini-profiler-0.9.8/lib/patches/db/pg.rb:90:in `exec'
```

I know ways around it easily enough, but it's kind of nice to do something like:

```
if can?(:manage, @thing)
...
else
...
end
```

What seems strange though is the following is fine:

```
[28] pry(main)> Thing.with_role(:admin, nil).pluck(:id)
   (0.6ms)  SELECT "things"."id" FROM "things" INNER JOIN "roles" ON "roles".resource_type IN ('Thing') AND
                                    ("roles".resource_id IS NULL OR "roles".resource_id = "things".id) WHERE ("roles".name IN ('admin') AND "roles".resource_type IN ('Thing'))
=> []
```

I have not had a chance to look at the code yet to submit a pull request and try and fix it. Otherwise, if there is a better way to do this that I'm not currently aware of I would love to know. 

Thanks,
Chris

**Comment by jotterlei** [comment]
@chriskaukis I came across this as well.  Had to do this format to get around it.  
`can :manage, Thing, id: user.new_record? ? [] : Thing.with_role(:admin, user).pluck(:id)`

Strange to think you and I are the only ones to see this yet.  Maybe the user.new_record? is the route others have taken

**Comment by bugzpodder** [comment]
Ran into this issue in a test as well.

**Comment by tom-lord** [comment]
Ran into this too. @jotterlei 's suggestion above solved the issue for me, although it feels wrong to need such a workaround.

**Comment by TylerRick** [comment]
Ran into this as well, after upgrading my app from a much older version of rolify (3.2.0 -> 5.2.0) and rails.

This used to work in rolify 3.2. Luckily I still had the old version of the app around and I could compare the queries generated. It used to generate this query:
```
... AND ("roles".id IN (NULL) AND ...
```
and simply return an empty result set. Now it generates this query:
```
... AND ("roles"."id" IN () AND ...
```
and results in a `PG::SyntaxError` as noted above.

It looks like we could pretty easily get it to go back to the old, safe behavior by simply passing an array instead of a relation to the `where('... IN (?) ...', roles)` condition.

Here's a really simple example you can try on the console to see the difference in behavior:

With non-empty array or relation:
```
main > Role.where('id in (?)', Role.last(1)).to_a
  Role Load (0.8ms)  SELECT  "roles".* FROM "roles" ORDER BY "roles"."id" DESC LIMIT 1
=> [#<Role:...]
```

With empty array:
```
main > Role.where('id in (?)', []).to_a
  Role Load (0.7ms)  SELECT "roles".* FROM "roles" WHERE (id in (NULL))
=> []
```

With empty relation:
```
main > Role.where('id in (?)', Role.none).to_a
  Role Load (0.9ms)  SELECT "roles".* FROM "roles" WHERE (id in ())
ActiveRecord::StatementInvalid: PG::SyntaxError: ERROR:  syntax error at or near ")"
```

Can we please change it back so that it's safe to call `Thing.with_role(:admin, user)` without risking getting an exception if user is not persisted?

**Comment by TylerRick** [comment]
Speaking of making things safe even if record is not yet persisted... we should probably also fix #476
